### PR TITLE
Unit tests for pallet EPM-MB

### PR DESCRIPTION
**Pull request by Zebedeusz**

### CODE REVIEW

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs [hunk 1]
```diff
@@ -324,10 +679,190 @@ mod solution_data_provider {
...
+
+			let score = ElectionScore { minimal_stake: 40, sum_stake: 5, sum_stake_squared: 0 };
+			assert_ok!(SignedPallet::register(RuntimeOrigin::signed(30), score));
+
+			assert_eq!(<SignedPallet as SolutionDataProvider>::get_score(), Some(higher_score));
+		})
+	}
+
+	#[test]
+	fn get_score_returns_entry_with_lowest_sum_stake_squared_given_entries_with_equal_minimal_stakes_and_sum_stakes(
```
Comment (gpestana) [inline substrate/frame/election-provider-multi-block/src/signed/tests.rs:700]: I would probably simplify the test names :) this is too long and too descriptive. Again I think this test could be merged in the above test, where all the permutations of scores and ordering are tested.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs [hunk 2]
```diff
@@ -324,10 +679,190 @@ mod solution_data_provider {
...
+			roll_to_phase(Phase::Signed);
+
+			assert_ok!(SignedPallet::register(origin.clone(), Default::default()));
+			assert_ok!(SignedPallet::submit_page(origin, 0, Some(Default::default())));
+
+			assert_eq!(<SignedPallet as SolutionDataProvider>::get_paged_solution(12345), None)
+		})
+	}
+
+	#[test]
```
Comment (gpestana) [inline substrate/frame/election-provider-multi-block/src/signed/tests.rs:744]: This test could be merged with the above test. Calling it `get_wrong_solution_page_works` and then testing all the cases where the None is returned due to bad page index is a good call !

**Review by gpestana — COMMENTED** [review]
A few comments and nits, especially around merging test cases that are similar and simplifying the test code.

Other than that, this is great, thanks for the help!

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs [hunk 3]
```diff
@@ -224,16 +231,364 @@ mod calls {
...
 
 	#[test]
-	fn force_clear_submission_works() {
+	fn register_and_submit_page_and_bail_prohibitted_in_phase_off() {
 		ExtBuilder::default().build_and_execute(|| {
-			// TODO
+			// let some_bn = 0;
+			// let some_page_index = 0;
+
+			let account_id = 99;
```
[the remark this answers is not in the document]
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:246]: I prefer to use a variable than a number that looks a little magical. This way wherever this variable is used, one can see what it actually represents.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs [hunk 4]
```diff
@@ -224,16 +231,364 @@ mod calls {
...
+			);
+
+			assert_err!(
+				SignedPallet::bail(RuntimeOrigin::signed(account_id)),
+				NotAcceptingSubmissions::<Runtime>,
+			);
+		})
+	}
+
+	#[test]
```
[the remark this answers is not in the document]
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:412]: I was rather testing what the test name says, so: given there are existing submissions, if someone bails from a submission that they didn't do, then the balances of all the participants, the one that had a submission and especially the one that bailed but didn't have one will remain unchanged.
Maybe it's too much of a defensive approach but I felt like it's worth having coverage there.
So that's my reasoning but if you don't like it here, then I'll remove it, sure.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs [hunk 5]
```diff
@@ -224,16 +231,364 @@ mod calls {
...
+				(100 - base_deposit - page_deposit, base_deposit + page_deposit)
+			);
+
+			// bailing account balance unchanged
+			assert_eq!(balances(bailing_account_id), (100, 0));
+		})
+	}
+
+	#[test]
+	fn force_clear_submission_fails_if_called_by_account_none() {
```
[the remark this answers is not in the document]
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:477]: This is a test for force_clear but I guess you mean the one for bail. 
Anyway, if I understand correctly you would put all the test cases for "bail" function in one test function.
If yes, then I would be against it. It would surely provide the coverage we need and verify all we need but it would result in overloaded tests, checking lots of things, instead of many tests, each checking very specific outcome.
So let me understand your reasoning here.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs [hunk 6]
```diff
@@ -224,16 +231,364 @@ mod calls {
...
+				assert_err!(
+					SignedPallet::force_clear_submission(RuntimeOrigin::root(), 0, account_id),
+					NotAcceptingSubmissions::<Runtime>,
+				);
+			}
+		})
+	}
+
+	#[test]
+	fn force_clear_submission_fails_if_submitter_done_no_submissions_at_all() {
```
[the remark this answers is not in the document]
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:514]: Is there a specific reason why you're suggesting the approach of having few tests with lots of checks inside each one?
My preferred approach is exactly the opposite one, so I'd like to understand your point of view.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs (hunk 1 above)
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:700]: Yeah, I got carried away with these names a little :)
I am very much for desciptive but true that this is too much. 
I'll work on it.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs (hunk 2 above)
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:744]: As above. I'm not in favour for merged scenarios so let's discuss that bigger topic first

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs [hunk 7]
```diff
@@ -224,16 +231,364 @@ mod calls {
...
+		ExtBuilder::default().build_and_execute(|| {
+			assert_err!(
+				SignedPallet::force_clear_submission(RuntimeOrigin::none(), 0, 99),
+				BadOrigin
+			);
+		})
+	}
+
+	#[test]
+	fn force_clear_submission_fails_if_called_in_phase_other_than_off() {
```
[the remark this answers is not in the document]
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:346]: Done

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (rockbmb) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:336]: Quick 2-cents from me: in this case, testing different origins is a simple procedure, so I get @Zebedeusz's point of having different `#[test]`s

If testing different origins required a (more or less) similar setup everytime, doing it all in the same test would make more sense as @gpestana suggests.

Example: a verbose [integration test](https://github.com/paritytech/polkadot-sdk/pull/6377/files#diff-8501c51f73a211c9a61241de6aa68de2e8f87d086c523fd52c7cddcfedec4d59R79) for Westend's people parachain checking for different wrong origins.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (gpestana) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:279]: > so: given there are existing submissions, if someone bails from a submission that they didn't do, then the balances of all the participants, the one that had a submission and especially the one that bailed but didn't have one will remain unchanged.

This can be tested using the `assert_noop!` macro. Basically, when `assert_noop!(call(), Error)`, the macro checks a couple things:
1. The `call()` will return the expected `Error
2. The Ext storage was not changed after calling `call()`. this means that the balances, and other states remain the same.

I'd just remove the whole set up of the registration etc and call an `assert_noop!(bail_for(1), Err)`. This tests that the bailing does not work in this case and the storage root remained the same, so no state has changed (+ simplify the tests considerably).

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs [hunk 8]
```diff
@@ -224,16 +231,225 @@ mod calls {
...
+				assert_err!(
+					SignedPallet::bail(RuntimeOrigin::signed(account_id)),
+					NotAcceptingSubmissions::<Runtime>,
+				);
+			}
+		})
+	}
+
+	#[test]
+	fn bail_while_having_no_submissions_does_not_modify_balances() {
```
Comment (gpestana) [inline substrate/frame/election-provider-multi-block/src/signed/tests.rs:280]: Same as above, `assert_noop!()` will simplify the test case, make it more comprehensive and simpler.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (gpestana) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:375]: The reasons why I'd merge these small tests cases are:
1. Simplify - we can re-use setup code across tests cases;
2. Context - the tests are basically testing the same unit (the force_clear) behaviour
3. Follow-along/readability - it's easier to undestand all the invariants of `force_clear` etc if they are all in the same test, we can use comments to explain what is being tested and why errors happen, etc
4. Low verbosity - especially due to 1.

See, for example, this test case in staking: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/staking/src/tests.rs#L1283

We test within a test all the flow of `Call::unbound`. We could split it in different tests but the setup/boilerplate code would be repeated all over the individual tests. Also through that test you can read and understand the whole unbound flow, requirements, invariants, etc

[the remark this answers is not in the document]
Comment (gpestana) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:375]: https://github.com/paritytech/polkadot-sdk/blob/master/substrate/frame/staking/src/tests.rs#L1495 being another good example, where `assert_noop!` is used too to ensure that errors do not change the storage root (and state).

File: substrate/frame/election-provider-multi-block/src/mock/mod.rs [hunk 9]
```diff
@@ -66,6 +67,14 @@ pub type T = Runtime;
 pub type Block = frame_system::mocking::MockBlock<Runtime>;
 pub(crate) type Solver = SequentialPhragmen<AccountId, sp_runtime::PerU16, ()>;
 
+pub struct Weighter;
+
+impl Get<RuntimeDbWeight> for Weighter {
```
[the remark this answers is not in the document]
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/mock/mod.rs:72]: Take a look at test called `on_initialize_returns_specific_weight_in_off_phase`. 
If DbWeight is set to () it fails.
So I defined it as such for the reason explained in the test.

File: substrate/frame/election-provider-multi-block/src/mock/mod.rs [hunk 10]
```diff
@@ -325,7 +336,12 @@ impl ExtBuilder {
 	}
 
 	pub(crate) fn desired_targets(self, desired: u32) -> Self {
-		DesiredTargets::set(desired);
+		DesiredTargets::set(Ok(desired));
+		self
+	}
+
+	pub(crate) fn no_desired_targets(self) -> Self {
```
[the remark this answers is not in the document]
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/mock/mod.rs:343]: In short to make some actions fail due to desired targets not being configured.
Needed for:
1. Test `worker_fails_to_mine_solution` 
2. Test `desired_targets_not_in_snapshot`- note: this one I will simplify a bit.

Are these not valid scenarios to test?
It seemed like valuable checks to me.

File: substrate/frame/election-provider-multi-block/src/signed/tests.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (Zebedeusz) [reply substrate/frame/election-provider-multi-block/src/signed/tests.rs:375]: I think our dispute boils down to the level at which we see these tests. 
You (and maybe it's the way in the whole repo, I can't know yet) advocate for a higher level tests, where a single test function would run a validation flow through the tested function with some failure checks alongside. This makes the test functionality test or flow test or smth like that. So it reads like a story.
My initial approach is in general to test the smallest possible units of functionality first then go to flow tests.
So each unit test function is called `fn_results_with_x_when_z` where conditions and expected results are clear in the name of the test. This way expectations for a given function are set in the unit tests.
Then I go for writing functional/flow tests and I do them as you would, just with less failure checks.

As to the verbosity and boilerplate, it can be often packed in setup functions (to avoid duplication) but it's rather good to be explicit in test setup so as to describe the initial conditions.

So to sum up, I think we actually align on the way to write tests, it's just that I went lower in the test pyramid, which I would argue is a good thing and to improve readability we could arrange the tests differently.
